topi/python/topi/cuda/rcnn/proposal.py

```diff
@@ -0,0 +1,355 @@
+            pred_y1 = tvm.max(tvm.min(pred_y1, im_height - 1.0), 0.0)
+            pred_x2 = tvm.max(tvm.min(pred_x2, im_width - 1.0), 0.0)
+            pred_y2 = tvm.max(tvm.min(pred_y2, im_height - 1.0), 0.0)
+
+            real_height = (im_height / feature_stride).astype('int32')
+            real_width = (im_width / feature_stride).astype('int32')
+
+            bbox_w = pred_x2 - pred_x1 + 1.0
+            bbox_h = pred_y2 - pred_y1 + 1.0
+            min_size = p_im_info[b * 3 + 2] * rpn_min_size
+
+            pred_score = p_score[((b * num_anchors * 2 + num_anchors + k) * height + h) * width + w]
+            pred_score = tvm.select(tvm.any(h >= real_height, w >= real_width), -1.0, pred_score)
+
+            p_out[out_index * 5 + 0] = pred_x1
+            p_out[out_index * 5 + 1] = pred_y1
+            p_out[out_index * 5 + 2] = pred_x2
+            p_out[out_index * 5 + 3] = pred_y2
+            p_out[out_index * 5 + 4] = pred_score
+
+            with ib.if_scope(tvm.any(bbox_w < min_size, bbox_h < min_size)):
+                p_out[out_index * 5 + 0] -= min_size / 2.0
+                p_out[out_index * 5 + 1] -= min_size / 2.0
+                p_out[out_index * 5 + 2] += min_size / 2.0
+                p_out[out_index * 5 + 3] += min_size / 2.0
+                p_out[out_index * 5 + 4] = -1.0
+
+    return ib.get()
+
+
+def argsort_ir(data_buf, out_index_buf):
+    """Batched odd-even transposition sort.
+
+    Parameters
+    ----------
+    data_buf : tvm.schedule.Buffer
+        2-D with shape [batch, num_bbox]
+
+    out_index_buf : tvm.schedule.Buffer
+        2-D with shape [batch, num_bbox]. Indices of data in sorted order.
+
+    Returns
+    -------
+    stmt : Stmt
+        The result IR statement.
+    """
+    batch, num_bbox = get_const_tuple(data_buf.shape)
+    max_threads = int(tvm.target.current_target(allow_none=False).max_num_threads)
+    tx = tvm.thread_axis("threadIdx.x")
+    bx = tvm.thread_axis("blockIdx.x")
+    ib = tvm.ir_builder.create()
+    temp_data = ib.allocate("float32", (1,), name="temp_data", scope="local")
+    temp_index = ib.allocate("int32", (1,), name="temp_index", scope="local")
+    p_data = ib.buffer_ptr(data_buf)
+    index_out = ib.buffer_ptr(out_index_buf)
+    nthread_tx = max_threads
+    nthread_bx = num_bbox // max_threads + 1
+    ib.scope_attr(tx, "thread_extent", nthread_tx)
+    ib.scope_attr(bx, "thread_extent", nthread_bx)
+    tid = bx * nthread_tx + tx
+
+    with ib.for_range(0, batch, for_type="unroll") as b:
+        start = b * num_bbox
+        with ib.if_scope(tid < num_bbox):
+            index_out[start + tid] = tid
+
+        with ib.for_range(0, num_bbox) as k:
+            with ib.if_scope(tid < (num_bbox + 1) // 2):
+                offset = start + 2 * tid + (k % 2)
+                with ib.if_scope(
+                    tvm.all(offset + 1 < num_bbox, p_data[offset] < p_data[offset + 1])):
+                    temp_data[0] = p_data[offset]
```

Comment:
Because different *offset*s are executed in parallel, if two different *offset*s both satisfy the condition, they'll compete the usage of temp_data[0] (same thing happens for temp_index[0]). In that case, the argsort result is wrong.

Reply:
`temp_data` is local scoped so each thread will have its own `temp_data`

Reply:
Have you tested your argsort on slightly larger dataset such as data_buf has shape (1, 500)? I tested it locally and it failed. Testing script can be found here: https://gist.github.com/Laurawly/66e8105c8db300bbce0771c1e58853ad

Reply:
This is a potential bug in tvm, `temp_index` and `temp_data` are in global memory, though I declared them to be local.

Reply:
hmm, it would be great if you can look a bit into it

Reply:
Can we import thrust as dependency since it difficult to write a efficient one with ir builder?

Reply:
We do need to fix the bug here first. The main benefit of using ir builder/hybrid script is that it can generalize to other GPU backends such as openCL and vulkan.

Reply:
@were TVM instrinsic `tvm_storage_sync` is not supported in hrbrid. This is needed to produce `__syncthreads` on CUDA

Reply:
@vinx13 just wondering, is there a way to implement sorting without global barrier? Other backends (opencl, vulkan, rocm) doesn't support global barrier or atomic at the moment. I want to port this to amdgpu later, but then I need to add support for atomic in our llvm backend (which I am not sure how to do).

Reply:
@vinx13 I tried using multiple kernels to write sorting to achieve synchronization effect when writing nms in SSD. But with the input data size getting larger, there can still be problems such as data conflicts. The best way is to see why tvm is not allocating local gpu memory but only global memory.

Reply:
@Laurawly tvm can correctly allocate local memory as long as you put the allocate statement inside the thread scope (scope_attr statement in ir builder). otherwise storage write pass cannot find the attach point of the allocation and put it to the beginning (which will be allocated as global, but we may add an assertion)

Reply:
@vinx13 I pulled the changes in #2473 and tested your argsort locally but it still fails on test cases with shape (1,500)

Reply:
@Laurawly Seems I somehow dropped some commits here, could you check again?

Reply:
@vinx13 Yeah, it works for me for up to (1, 6000). But it's good enough for the test cases.